src/funk/fd_funk_txn.c

```diff
@@ -0,0 +1,723 @@
+   in txn_idx's family.  Callers have already validated our input
+   argumnets.  The caller should validate the return index. */
+
+static inline ulong
+fd_funk_txn_oldest_sibling( fd_funk_t *     funk,
+                            fd_funk_txn_t * map,
+                            ulong           txn_max,
+                            ulong           txn_idx ) {
+
+  ulong parent_idx = fd_funk_txn_idx( map[ txn_idx ].parent_cidx );
+
+  if( FD_LIKELY( fd_funk_txn_idx_is_null( parent_idx ) ) ) return fd_funk_txn_idx( funk->child_head_cidx ); /* opt for incr pub */
+
+  ASSERT_IN_PREP( parent_idx );
+
+  return fd_funk_txn_idx( map[ parent_idx ].child_head_cidx );
+}
+
+/* fd_funk_txn_cancel_sibling_list cancels siblings from sibling_idx down
+   to the youngest sibling inclusive in the order from youngest to
+   sibling_idx.  Callers have already validated our input arguments
+   except sibling_idx.  Returns the number of cancelled transactions
+   (should be at least 1).  If any sibling is skip_idx, it will be not
+   be cancelled but still tagged as visited.  Passing
+   FD_FUNK_TXN_IDX_NULL for skip_idx will cancel all siblings from
+   sibling_idx to the youngest inclusive. */
+
+static ulong
+fd_funk_txn_cancel_sibling_list( fd_funk_t *     funk,
+                                 fd_funk_txn_t * map,
+                                 ulong           txn_max,
+                                 ulong           tag,
+                                 ulong           sibling_idx,
+                                 ulong           skip_idx ) {
+
+  ulong cancel_stack_idx = FD_FUNK_TXN_IDX_NULL;
+
+  /* Push siblings_idx and its younger siblings inclusive (skipping
+     sibling skip_idx if encounter) onto the cancel stack from oldest to
+     youngest (such that we cancel youngest to oldest). */
+
+  for(;;) {
+
+    /* At this point, sibling_idx is a sibling we might want to add to
+       the sibling stack.  Validate and tag it. */
+
+    ASSERT_IN_PREP ( sibling_idx );
+    ASSERT_UNTAGGED( sibling_idx );
+    map[ sibling_idx ].tag = tag;
+
+    if( FD_UNLIKELY( sibling_idx!=skip_idx ) ) { /* Not skip_idx so push onto the cancel stack, opt for non-compete */
+      map[ sibling_idx ].stack_cidx = fd_funk_txn_cidx( cancel_stack_idx );
+      cancel_stack_idx = sibling_idx;
+    }
+
+    ulong younger_idx = fd_funk_txn_idx( map[ sibling_idx ].sibling_next_cidx );
+    if( FD_LIKELY( fd_funk_txn_idx_is_null( younger_idx ) ) ) break; /* opt for non-compete */
+    sibling_idx = younger_idx;
+
+  }
+
+  /* Cancel all transactions and their descendants on the cancel stack */
+
+  ulong cancel_cnt = 0UL;
+
+  while( !fd_funk_txn_idx_is_null( cancel_stack_idx ) ) { /* TODO: peel first iter to make more predictable? */
+    ulong sibling_idx = cancel_stack_idx;
+    cancel_stack_idx  = fd_funk_txn_idx( map[ sibling_idx ].stack_cidx );
+
+    cancel_cnt += fd_funk_txn_cancel_family( funk, map, txn_max, tag, sibling_idx );
```

Comment:
Seems like you are using the same stack for this function as well as the interior of cancel_family. Is this correct?

Reply:
Yep ... saves space and it is a security measure as the families of two different siblings should never share a member.  So this allows the ASSERT_UNTAGGED stuff to detect if there are any such collisions between siblings and their families.